pkgs/applications/networking/browsers/ladybird/default.nix

```
@@ -0,0 +1,92 @@
+{ lib
+, gcc11Stdenv
+, fetchFromGitHub
+, cmake
+, ninja
+, unzip
+, wrapQtAppsHook
+, makeWrapper
+, qtbase
+, qttools
+}:
+
+let serenity = fetchFromGitHub {
+  owner = "awesomekling";
+  repo = "serenity";
+  rev = "094ba6525f0217f3b8d5e467cef326caeb659e8a";
+  hash = "sha256-IHXe2Td9iRSL1oQVwL2gZHxEM2ID4SghZwK6ewjFV1Y=";
+};
+
+in gcc11Stdenv.mkDerivation {
+  pname = "ladybird";
+  version = "unstable-2022-07-20";
+
+  # Remember to update `serenity` too!
+  src = fetchFromGitHub {
+    owner = "awesomekling";
+    repo = "ladybird";
+    rev = "9e3a1f47d484cee6f23c4dae6c51750af155a8fc";
+    hash = "sha256-1cPWpPvjM/VcVUEf2k+MvGvTgZ3Fc4LFHZCLh1wU78Y=";
+  };
+
+  nativeBuildInputs = [
+    cmake
+    ninja
+    unzip
+    wrapQtAppsHook
+    makeWrapper
+  ];
+
+  buildInputs = [
+    qtbase
+  ];
+
+  cmakeFlags = [
+    "-DSERENITY_SOURCE_DIR=${serenity}"
+    # Disable network operations
+    "-DENABLE_TIME_ZONE_DATABASE_DOWNLOAD=false"
+    "-DENABLE_UNICODE_DATABASE_DOWNLOAD=false"
+  ];
+
+  NIX_CFLAGS_COMPILE = [ "-Wno-error" ];
+
+  installPhase = ''
+    runHook preInstall
+    install -Dm755 ladybird $out/bin/ladybird
+    mkdir -p $out/lib/ladybird
+    cp -d _deps/lagom-build/*.so* $out/lib/ladybird/
+    runHook postInstall
+  '';
+
+  # Patch rpaths
+  preFixup = ''
+    for f in $out/bin/ladybird $out/lib/ladybird/*.so; do
+      old_rpath=$(patchelf --print-rpath "$f")
+      # Remove reference to libraries from build directory
+      rpath_without_build=$(sed -e 's@[^:]*/_deps/lagom-build:@@g' <<< $old_rpath)
+      # Add directory where we install those libraries
+      new_rpath=$out/lib/ladybird:$rpath_without_build
+      patchelf --set-rpath "$new_rpath" "$f"
+    done
+  '';
+
+  # According to the readme, the program needs access to the serenity sources
+  # at runtime
+  postFixup = ''
+    wrapProgram $out/bin/ladybird --set SERENITY_SOURCE_DIR "${serenity}"
+  '';
+
+  # Stripping results in a symbol lookup error
+  dontStrip = true;
+
+  meta = with lib; {
+    description = "A browser using the SerenityOS LibWeb engine with a Qt GUI";
+    homepage = "https://github.com/awesomekling/ladybird";
+    license = licenses.bsd2;
+    maintainers = with maintainers; [ fgaz ];
+    # SerenityOS only works on x86, and can only be built on unix systems
+    platforms = intersectLists platforms.unix (platforms.x86_64 ++ platforms.i686);
+    # We use patchelf in preFixup
+    broken = !gcc11Stdenv.isLinux;
```

```suggestion
    platforms = [ "x86_64-linux" ];
```
effectively this translates to `x86_64-linux`

---

It doesn't:

```
nix-repl> with (import <nixpkgs> {}).lib; intersectLists platforms.unix (platforms.x86_64 ++ platforms.i686)
[ "x86_64-cygwin" "x86_64-darwin" "x86_64-freebsd" "x86_64-solaris" "x86_64-linux" "x86_64-netbsd" "x86_64-openbsd" "x86_64-redox" "i686-cygwin" "i686-darwin" "i686-freebsd" "i686-linux" "i686-netbsd" "i686-openbsd" ]
```

Furthermore, `platforms` and `broken` have different meanings.

---

cygwin, freebsd, solaris, netbsd, openbsd, redox is normally ignored when choosing meta.platforms.



> edit: even if we coalesce `platform` and `broken`:

oh yeah, then it would be ``platforms = platforms.linux``. Just saying we should keep things simple and not set broken if it never worked before.

---

Done. I used `[ "x86_64-linux" "i686-linux" ]`

---

Just an FYI ladybird has been built on aarch64 and Android too without _much_ fuss, so it's not specially i686 and x86_64 (though SerenityOS itself is for the time being, though there's work on an aarch64)

---

Cross compiling take some ... CMake massaging though considering the host tools are in the same project().  Given the unstable tag, I'm guessing fgaz is intending to revisit this periodically if we improve the build.

---

@MacDue good to know! I already tried to build on aarch64 and it failed (can't remember the error). I'll retry periodically and update `platforms` when it builds

---

By the way, out of curiosity, how did you find this pr? I didn't advertise it on any serenityos channel yet

---

Here's the error:

```
[487/2587] Linking CXX executable _deps/lagom-build/Tools/CodeGenerators/LibWeb/GenerateCSSTransformFunctions
FAILED: _deps/lagom-build/Tools/CodeGenerators/LibWeb/GenerateCSSTransformFunctions 
: && /nix/store/pa9hmkgsfwsl9rpmk37fqimi8h8spnqj-gcc-wrapper-11.3.0/bin/g++ -O3 -DNDEBUG  _deps/lagom-build/Tools/CodeGenerators/LibWeb/CMakeFiles/GenerateCSSTransformFunctions.dir/GenerateCSSTransformFunctions.cpp.o -o _deps/lagom-build/Tools/CodeGenerators/LibWeb/GenerateCSSTransformFunctions  -Wl,-rpath,/build/source/build/_deps/lagom-build::::::::::::::::::::::::::::::::::::::::::::::::::::::::::  _deps/lagom-build/liblagom-main.so.0.0.0  _deps/lagom-build/liblagom-core.so.0.0.0  -lcrypt && :
/nix/store/ld11fgq92p6bci267ywy55j9825sbi66-binutils-2.38/bin/ld: _deps/lagom-build/Tools/CodeGenerators/LibWeb/GenerateCSSTransformFunctions: hidden symbol `__aarch64_ldadd2_acq_rel' in /nix/store/ln8312k088kl7psd0jc33x7p2ja62lrw-gcc-11.3.0/lib/gcc/aarch64-unknown-linux-gnu/11.3.0/libgcc.a(ldadd_2_4.o) is referenced by DSO
/nix/store/ld11fgq92p6bci267ywy55j9825sbi66-binutils-2.38/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
```

Full log: 
[ladybird-aarch64-log.txt](https://github.com/NixOS/nixpkgs/files/9183932/ladybird-aarch64-log.txt)

---

Interesting. It might need -latomic or so.

Hah, someone posted the PR in a serenity discord channel, I'm not sure how they came across it.

---

It needs to be explicitly linked to libatomic. AArch64 doesn't have all the atomic instructions that x86-64 has, so some operations have to be libcalls.